### ISSUE TITLE
Replace DGL PRNG with a better one

### ISSUE THREAD

**Issue by mfbalin** [dmlc/dgl#4785]
## 🚀 Feature

Replace std::default_random_engine with a better proven PRNG that passes statistical tests.

## Motivation

I noticed that the DGL PRNG is just STL default random engine, which is just std::minstd_rand0. The set seed function sets the seed of the PRNG by adding the thread id to the seed. But since the minstd_rand0 is not a good quality PRNG, two threads, one with id 0 and one with id 1 for example will have seeds, seed and seed+1. The sequence of numbers generated by these two threads will be correlated with each other, they wouldn't pass the randomness tests available in the literature. I would recommend replacing the underlying random number engine with a better one, pcg is one that DGL could consider. You may ask why this is important, it is important because an experimental result could be worse than it potentially could be silently just because of the bad quality of the PRNG.

More modern PRNGs such as pcg has the capability of having streams, so that different threads can be initialized with same seed, but on a different stream. https://www.pcg-random.org/useful-features.html

Numpy for example has switched to using pcg as their default PRNG for these reasons. The new PRNG doesn't have to be PCG, there are alternatives to it. I just wanted to start a discussion about this topic. https://numpy.org/doc/stable/reference/random/upgrading-pcg64.html

**Comment by mufeili** [comment]
It seems that you added that code snippet. Any thoughts? @classicsong

**Comment by BarclayII** [comment]
> You may ask why this is important, it is important because an experimental result could be worse than it potentially could be silently just because of the bad quality of the PRNG.

I also mentioned in the PR #4668 ; it's better to provide evidence on this.  For instance, how much would the accuracy vary by switching the RNG alone?

**Comment by mfbalin** [comment]
My use case requires using the PRNG as a function of a given random seed and a given integer t, and I expect the resulting uniform random numbers to be uncorrelated with each other for any given pair of random_seed and the integer t.

With pcg32 number of sampled vertices and edges: I seed PRNG with random seed and skip ahead t times with it.
`NS_1024_10 [(215705, 204070, 2034612), (204070, 167180, 1669726), (167180, 69542, 694811), (69542, 10331, 103009), (10331, 1024, 10023)]
LABOR-0_1024_10 [(81841, 57935, 568844), (57935, 35431, 362985), (35431, 17856, 171143), (17856, 6835, 69360), (6835, 1024, 10057)]
LABOR-1_1024_10 [(56803, 40971, 709630), (40971, 26282, 473390), (26282, 14512, 252447), (14512, 6263, 115211), (6263, 1024, 18086)]
LABOR-*_1024_10 [(49241, 36118, 2707147), (36118, 23952, 1796223), (23952, 13743, 1029221), (13743, 6087, 407911), (6087, 1024, 29988)]`

With std::default_random_engine that DGL uses, https://github.com/mfbalin/dgl-1/tree/dgl_prng_labor, I seed the PRNG with random_seed ^ t:
`NS_1024_10 [(215664, 204033, 2033973), (204033, 167294, 1670759), (167294, 69320, 692594), (69320, 10315, 102837), (10315, 1024, 10003)]
LABOR-0_1024_10 [(23111, 19144, 253396), (19144, 13659, 213376), (13659, 8657, 169905), (8657, 4494, 105825), (4494, 1024, 5574)]
LABOR-1_1024_10 [(9015, 8641, 297526), (8641, 7892, 285135), (7892, 6510, 252143), (6510, 4141, 168423), (4141, 1024, 8339)]
LABOR-*_1024_10 [(13000, 12601, 902670), (12601, 11704, 801585), (11704, 9648, 625005), (9648, 4999, 310638), (4999, 1024, 18627)]`

**Comment by mfbalin** [comment]
I know that using std::default_random_engine in this way, I am not expected to get good results. But this is just an example showing how correlated the numbers sampled from different random_seeds can be.

**Comment by mfbalin** [comment]
If the current implementation is used in the future, there is the risk of some experiments not getting as good results as it could, and there is almost no way anybody will be able to debug it and find out that the multithreaded use of the PRNG could be the problem.

**Comment by BarclayII** [comment]
> If the current implementation is used in the future, there is the risk of some experiments not getting as good results as it could, and there is almost no way anybody will be able to debug it and find out that the multithreaded use of the PRNG could be the problem.

I understand that using STL PRNG would have correlation.  The question is, does such correlation have an impact on GNN training.  So far I have yet to see evidence supporting or refuting this hypothesis.

**Comment by mfbalin** [comment]
It is upto you. Algorithms like stochastic gradient descent depend on the fact that the PRNG is sufficiently random. I just wanted to point out a problem I saw, so that DGL gets better for everyone. Maybe you will reconsider in the future. In order to prove or disprove the PRNG is a problem in the way you want, one would have to do multithreaded training that uses the DGL PRNG somehow, maybe thousands of times, record the accuracies, compute the mean and standard deviation of the mean, and see if the mean of the current code and the fixed code are equal by doing a p-value analysis, which I am not going to be doing. I bet there will be a provable difference.

**Comment by mfbalin** [comment]
If you don't want to replace the PRNG, maybe you can take a look at the quick fix in #4808.

**Comment by BarclayII** [comment]
> It is upto you. Algorithms like stochastic gradient descent depend on the fact that the PRNG is sufficiently random, their theory only holds when randomness is assumed.

I agree with this.  However in practice, strict statistical randomness may not bring substantial benefit comparing to "some sort of randomness".  For instance, PyTorch uses [MT19937](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/CPUGeneratorImpl.h#L29), which is known to be [problematic](http://dirsig.cis.rit.edu/docs/new/random_numbers.html) comparing to PCG, but I didn't see they have any issue with it.  In fact, they uses MT19937 for sampling minibatches in SGD (more precisely, they use a normal CPU PRNG in their DataLoader, which is MT19937).  So I guess using a not-so-random RNG is fine in general.

**Comment by mfbalin** [comment]
MT19937 is a much better choice in terms of randomness compared to std::minstd_rand0, see the [table](https://arvid.io/2018/06/30/on-cxx-random-number-generator-quality/).

**Comment by BarclayII** [comment]
So would MT19937 be a more in-the-middle replacement of both PCG (which is slow) and `default_random_engine` (which doesn't produce good random numbers)?

**Comment by mfbalin** [comment]
Benchmarks for prngs can be found here, I am sure there are more benchmarks available. https://github.com/lemire/testingRNG

Pcg is better than mt19937 in every way, so there would be no reason to switch to it. Pcg is the same as default random engine

**Comment by yaox12** [comment]
Fixed by https://github.com/dmlc/dgl/pull/4807.
